File: leetcode/go/107.binary-tree-level-order-traversal-ii.go
/*
 * @lc app=leetcode id=107 lang=golang
 *
 * [107] Binary Tree Level Order Traversal II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
	var ret [][]int
	cur_level := 0
	parent_nodes := list.New()
	child_nodes := list.New()
	if root == nil {
		return ret
	}
	parent_nodes.PushBack(root)
	for parent_nodes.Len() > 0 {
		cur_level++
		var level_data []int
		loop_node := parent_nodes.Front()
		for loop_node != nil {
			node := loop_node.Value.(*TreeNode)
			level_data = append(level_data, node.Val)
			if node.Left != nil {
				child_nodes.PushBack(node.Left)
			}
			if node.Right != nil {
				child_nodes.PushBack(node.Right)
			}
			loop_node = loop_node.Next()
		}
		ret = append([][]int{level_data}, ret...)
		parent_nodes.Init()
		parent_nodes.PushBackList(child_nodes)
		child_nodes.Init()
	}
	return ret
}

// @lc code=end

File: leetcode/cpp/107.binary-tree-level-order-traversal-ii.cpp
/*
 * @lc app=leetcode id=107 lang=cpp
 *
 * [107] Binary Tree Level Order Traversal II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ret;
        if (root == nullptr)
        {
            return ret;
        }
        std::queue<std::pair<TreeNode *, int>> data_queue;
        data_queue.push(std::make_pair(root, 0));
        while (!data_queue.empty())
        {
            auto item = data_queue.front();
            data_queue.pop();
           if (item.second == ret.size())
                ret.insert(ret.begin(), vector<int>());
            ret[0].push_back(item.first->val); 
            if (item.first->left != nullptr)
            {
                data_queue.push(make_pair(item.first->left, item.second + 1));
            }
            if (item.first->right != nullptr)
            {
                data_queue.push(make_pair(item.first->right, item.second + 1));
            }
        }
        return ret;
    }
};
// @lc code=end

